### ISSUE TITLE
“This condition will always return true” is, in fact, false

### ISSUE THREAD

**Issue by martynchamberlin** [microsoft/TypeScript#61531]
### 🔎 Search Terms

This condition will always return true since this 'Promise<any>' is always defined.

### 🕗 Version & Regression Information

- This changed between versions 4.2.3 and 4.3.5. Specifically around this part of the release: https://devblogs.microsoft.com/typescript/announcing-typescript-4-3/#always-truthy-promise-checks 

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=4.3.5#code/DYUwLgBATiDOCuwwC4IAUoHsC2BLWIAPAIYB2AngHwQA+EpiwEAvPSAO7pZ4EAUMsTMABuIFtQDeAKAiyIBMABVc2EJnhhevAJTiI0uYehxEkVg2DAZR2QKGjeF4NuuyAvgBoIARgAMvlzcXKWJYclIAYwgAM3hIsFxMUmg4nX1XCFxoiH4TJF0DGwhidmJcSAFTDIzDLJzK-PSiowikwVAAOmBMAHNeAHIwKA0AC3J+l2a3CBBgAibmuVbSdpAu3oHo4jmQccmitwzDw6lh0h0gA

### 💻 Code

```ts
let result: Promise<any> | null = new Promise(resolve => {
    setTimeout(() => {
        result = null
        resolve(null)
    }, 100)
})

async function run() {
    if (result) {
        await result
    
        if (result) {
            console.log('truthy')
        } else {
            console.log('falsey')
        }
    }
}

run()
```

### 🙁 Actual behavior

TS throws an error on line 12 which it shouldn't be throwing: 

> This condition will always return true since this 'Promise<any>' is always defined.(2801)
input.tsx(12, 13): Did you forget to use 'await'?

### 🙂 Expected behavior

TS should not throw this error on line 12: 

> This condition will always return true since this 'Promise<number>' is always defined.

It's making an opinion about a value at runtime that it doesn't have adequate context to adjudicate. 

### Additional information about the issue

That broader scope conditional on line 9 is narrowing the `result` type down from `Promise<any> | null` to `Promise<any>`, but then the check on line 12 is failing to account for the fact that `result` could get its value reassigned in virtue of the `await` on line 10. That `await` should cause the `result` type to expand back to the original `Promise<any> | null`, and make subsequent truthy checks allowable without type errors.

**Comment by MartinJohns** [comment]
Another duplicate of #9998. There's even a dedicated issue template for this: https://github.com/microsoft/TypeScript/blob/main/.github/ISSUE_TEMPLATE/types-not-correct-in-with-callback.md?plain=1
